core/domain/story_services.py

```diff
@@ -241,6 +246,226 @@ def save_new_story(committer_id, story):
+                    story.update_node_outline(change.node_id, change.new_value)
+                elif (change.property_name ==
+                      story_domain.STORY_NODE_PROPERTY_ACQUIRED_SKILL_IDS):
+                    story.update_node_acquired_skill_ids(
+                        change.node_id, change.new_value)
+                elif (change.property_name ==
+                      story_domain.STORY_NODE_PROPERTY_PREREQUISITE_SKILL_IDS):
+                    story.update_node_prerequisite_skill_ids(
+                        change.node_id, change.new_value)
+                elif (change.property_name ==
+                      story_domain.STORY_NODE_PROPERTY_DESTINATION_NODE_IDS):
+                    story.update_node_destination_node_ids(
+                        change.node_id, change.new_value)
+                elif (change.property_name ==
+                      story_domain.STORY_NODE_PROPERTY_EXPLORATION_ID):
+                    story.update_node_exploration_id(
+                        change.node_id, change.new_value)
+            elif change.cmd == story_domain.CMD_UPDATE_STORY_PROPERTY:
+                if (change.property_name ==
+                        story_domain.STORY_PROPERTY_TITLE):
+                    story.update_title(change.new_value)
+                elif (change.property_name ==
+                      story_domain.STORY_PROPERTY_DESCRIPTION):
+                    story.update_description(change.new_value)
+                elif (change.property_name ==
+                      story_domain.STORY_PROPERTY_NOTES):
+                    story.update_notes(change.new_value)
+                elif (change.property_name ==
+                      story_domain.STORY_PROPERTY_LANGUAGE_CODE):
+                    story.update_language_code(change.new_value)
+            elif (
+                    change.cmd ==
+                    story_domain.CMD_MIGRATE_SCHEMA_TO_LATEST_VERSION):
+                # Loading the story model from the datastore into a
+                # Story domain object automatically converts it to use the
+                # latest schema version. As a result, simply resaving the
+                # story is sufficient to apply the schema migration.
+                continue
+        return story
+
+    except Exception as e:
+        logging.error(
+            '%s %s %s %s' % (
+                e.__class__.__name__, e, story_id, change_list)
+        )
+        raise
+
+
+def _save_story(committer_id, story, commit_message, change_list):
+    """Validates a story and commits it to persistent storage. If
+    successful, increments the version number of the incoming story domain
+    object by 1.
+
+    Args:
+        committer_id: str. ID of the given committer.
+        story: Story. The story domain object to be saved.
+        commit_message: str. The commit message.
+        change_list: list(dict). List of changes applied to a story. Each
```

Comment:
Do you think we could have this be list(StoryChange) instead? In general, better to work with domain objects rather than dicts. (I.e. convert to domain objects in the controller and pass the domain objects around.) Ditto for other functions in this file that use change dicts (and elsewhere in the codebase if you like, e.g. for explorations, but that's optional).

Reply:
Yes, if you are passing StoryChange domain objects, you don't need the dict validation. Instead you could have all the domain objects (StoryChange etc.) subclass BaseChange and check instead that the list contains a list of BaseChange objects.

And then you don't need to do any domain object conversions in apply_change_list, too.

(I feel like, since this pattern is becoming more prevalent with all the new models, we should probably do it properly rather than replicate the current suboptimal structure of passing dict blobs around.)